### PR TITLE
feat(meta_contract): batch create eth accounts

### PR DESCRIPTION
**Pull request by zeroqn**
## New method for meta_contract: `MSG_BATCH_CREATE_ACCOUNTS`

https://github.com/nervosnetwork/godwoken-scripts/blob/5d282a632a308f5190171afa49a89a52266c2f5b/c/contracts/meta_contract.c#L102-L159

### CODE REVIEW

**Review by Flouse — APPROVED** [review]

**Review by jjyr — APPROVED** [review]

File: c/contracts/meta_contract.c [hunk 1]
```diff
@@ -97,6 +99,64 @@ int main() {
...
+    /* charge fee */
+    mol_seg_t fee_seg = MolReader_BatchCreateEthAccounts_get_fee(&msg.seg);
+    mol_seg_t amount_seg = MolReader_Fee_get_amount(&fee_seg);
+    mol_seg_t reg_id_seg = MolReader_Fee_get_registry_id(&fee_seg);
+
+    uint256_t fee_amount = {0};
+    _gw_fast_memcpy((uint8_t *)(&fee_amount), (uint8_t *)amount_seg.ptr,
+                    sizeof(uint128_t));
+
+    uint32_t reg_id = *(uint32_t *)reg_id_seg.ptr;
```
Comment (sopium) [inline c/contracts/meta_contract.c:112]: Potentially unaligned pointer cast and load? Use something like https://github.com/nervosnetwork/molecule/blob/5f94aa86d523d60e498ffefe01001fcf1ae70851/bindings/c/include/molecule_reader.h#L84-L99 instead?

Comment (jjyr) [reply c/contracts/meta_contract.c:112]: Another approach is `memcpy`, but we already use too many cast in the C code.

Comment (zeroqn) [reply c/contracts/meta_contract.c:112]: Good catch. I think we could add `-fsanitize=undefined`, we need a new toolchain image.

**Review by sopium — APPROVED** [review]

File: c/contracts/meta_contract.c [hunk 2]
```diff
@@ -109,7 +109,10 @@ int main() {
     _gw_fast_memcpy((uint8_t *)(&fee_amount), (uint8_t *)amount_seg.ptr,
                     sizeof(uint128_t));
 
-    uint32_t reg_id = *(uint32_t *)reg_id_seg.ptr;
```
Comment (Flouse) [inline c/contracts/meta_contract.c]: In `c/contracts` dir, there are some other places used this kind of scheme.


```C
c/contracts/eth_addr_reg.c:
  108      mol_seg_t reg_id_seg = MolReader_Fee_get_registry_id(&fee_seg);
  109:     uint32_t reg_id = *(uint32_t *)reg_id_seg.ptr;
  110  

  141      mol_seg_t reg_id_seg = MolReader_Fee_get_registry_id(&fee_seg);
  142:     uint32_t reg_id = *(uint32_t *)reg_id_seg.ptr;
  143  

c/contracts/meta_contract.c:
   81  
   82:     uint32_t reg_id = *(uint32_t *)reg_id_seg.ptr;
   83      ret = handle_fee(&ctx, reg_id, fee_amount);

  111  
  112:     uint32_t reg_id = *(uint32_t *)reg_id_seg.ptr;
  113      ret = handle_fee(&ctx, reg_id, fee_amount);

c/contracts/sudt.c:
  91  
  92:     uint32_t reg_id = *(uint32_t *)fee_reg_seg.ptr;
  93      uint32_t from_id = ctx.transaction_context.from_id;
```

Comment (zeroqn) [reply c/contracts/meta_contract.c]: We can fix these in different pr.

**Pull request merged**